File: 백준/Gold/17298. 오큰수/오큰수.cc
#include <iostream>
#include <stack>
using namespace  std;
int n, a[1000004];
stack<pair<int, int>> stk; // {value, idx}
int main(){
	ios_base::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);

	cin >> n;

	for(int i = 0; i < n; i++){
		int value;
		cin >> value;
		
		if(stk.size()){ // O
			if(stk.top().first < value){
				
				while(true){
					if(stk.size()){
						if(stk.top().first < value){
							a[stk.top().second] = value;
							stk.pop();
						}else break;
					}else break;
				}

				stk.push({value, i});
			}else{
				stk.push({value, i});
			}
		}else{ // X
			stk.push({value, i});
		}
	}

	while(!stk.empty()){
		a[stk.top().second] = -1;
		stk.pop();
	}

	for(int i = 0; i < n; i++){
		cout << a[i] << ' ';
	}
	return 0;
}

File: 백준/Gold/17298. 오큰수/오큰수.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main{
    static StringBuilder sb = new StringBuilder();
    static int n;
    static int[] a, answer;
    static Stack<Integer> stk = new Stack<>();// idx 넣기
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        a = new int[n];
        answer = new int[n];
        for(int i = 0; i < n; i++){
            answer[i] = -1;
        }
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < n; i++){
            a[i] = Integer.parseInt(st.nextToken());
            while(!stk.isEmpty() && a[stk.peek()] < a[i]){// (있냐 없냐 && a[stk.top()] < a[now])
                int idx = stk.peek();
                stk.pop();
                answer[idx] = a[i];
            }
            stk.add(i);// idx 넣기
        }
        for(int i = 0; i < n; i++){
            sb.append(answer[i] + " ");
        }
        System.out.println(sb);
    }
}